### PR TITLE
[fix] (training): correct gradient accumulation boundaries

### LINKED ISSUE #395: [Bug] Gradient accumulation only uses the last microbatch when using Accelerate accumulation
## Bug

When `gradient_accumulation_steps > 1`, the training loop calls `optimizer.zero_grad()` at the
beginning of the `accelerator.accumulate(model)` block.

With Accelerate, `optimizer.zero_grad()` only takes effect on the sync microbatch. As a
result, previous accumulated microbatch gradients are cleared right before the final backward
pass, so the optimizer update only uses the last microbatch.

## Impact

For example, with `gradient_accumulation_steps=4`, microbatches 1-3 run forward/backward but
their gradients are discarded on microbatch 4. The effective batch size is not increased as
intended, and the remaining gradient may also be scaled by the accumulation factor.

## Expected behavior

Gradients from all accumulation microbatches should contribute to one optimizer update.

## Suggested fix

Move `optimizer.zero_grad()` to after `optimizer.step()`, and only run gradient clipping /
logging / scheduler update on `accelerator.sync_gradients`.

## Environment

- Accelerate training
- `gradient_accumulation_steps > 1`
- Observed in the StarVLA training loop

### PR DESCRIPTION
**Pull request by mindorigin150**
## Description

Correct gradient accumulation behavior across the VLA, co-training, and VLM training entrypoints for both native Accelerate and DeepSpeed execution.

Optimizer-coupled work now runs only on real optimizer update boundaries, rather than once per microbatch.

## Motivation

Fixes #395.

Related to #86, #366, and #408.

The configured `gradient_accumulation_steps` was not consistently propagated to `Accelerator`, and several training paths coupled clipping, scheduler updates, progress/logging, evaluation, or checkpointing to microbatches instead of effective optimizer steps.

## Changes

- Centralize `Accelerator` construction and pass the wrapped trainer config's `gradient_accumulation_steps` through all three training entrypoints.
- Correct the native Accelerate path so gradients are cleared after an optimizer step and clipping/scheduler updates happen only when `sync_gradients` is true.
- Make VLA, co-training, and VLM DeepSpeed paths use the engine's `backward()`, accumulation-boundary signal, and `step()`; advance the user scheduler only at a real boundary.
- Select the DeepSpeed path through the explicit `DistributedType.DEEPSPEED` backend contract.
- Gate progress, logging, evaluation, and checkpointing on the explicit optimizer-step result.
- Set DeepSpeed `gradient_accumulation_steps` to `"auto"` so Accelerate supplies the trainer configuration.
- Add deterministic CPU regression tests and real DeepSpeed ZeRO-2/ZeRO-3 smoke tests for all three trainers.

## Testing

- [x] CPU configuration/wiring regression tests: 2 passed
- [x] CPU accumulation and optimizer-boundary behavior tests: 4 passed
- [x] Existing single-process distributed-safety tests: 4 passed
- [x] 1-GPU DeepSpeed matrix: VLA, VLM, and co-training × ZeRO-2 and ZeRO-3 passed
- [x] 2-GPU DeepSpeed matrix: VLA, VLM, and co-training × ZeRO-2 and ZeRO-3 passed
- [ ] Local full-model training for N steps without errors — not run; deterministic production-path tests isolate the affected two-microbatch updates
- [ ] Benchmark evaluation results — not applicable; this bug fix does not change model or benchmark semantics

| Benchmark | Metric | Result |
|-----------|--------|--------|
| N/A | N/A | No benchmark change |

- **Checkpoint**: N/A (no framework/benchmark integration)
- **Config**: `starVLA/config/deepseeds/ds_config.yaml`, `starVLA/config/deepseeds/zero3.yaml`
- **Environment**: 2× NVIDIA A100 80GB; Python 3.10.20; PyTorch 2.6.0+cu124; Accelerate 1.5.2; DeepSpeed 0.16.9
- **Reproduce**:

```bash
PYTHONPATH=. WANDB_MODE=disabled CUDA_VISIBLE_DEVICES= python tests/test_gradient_accumulation_config.py
PYTHONPATH=. WANDB_MODE=disabled CUDA_VISIBLE_DEVICES= python tests/test_training_gradient_accumulation.py
PYTHONPATH=. WANDB_MODE=disabled CUDA_VISIBLE_DEVICES= python tests/test_single_process_dist_safety.py

port=29910
for zero in 2 3; do
  for ranks in 1 2; do
    if [ "$ranks" -eq 1 ]; then gpus=0; else gpus=0,1; fi
    for trainer in vla vlm cotrain; do
      PYTHONPATH=. CUDA_VISIBLE_DEVICES=$gpus accelerate launch \
        --config_file starVLA/config/deepseeds/deepspeed_zero${zero}.yaml \
        --num_processes "$ranks" \
        --main_process_port "$port" \
        tests/test_training_gradient_accumulation_deepspeed.py \
        --trainer "$trainer"
      port=$((port + 1))
    done
  done
done
```

Additional checks:

- Modified/new test files pass Black and Ruff.
- Production lines added by this PR have no Ruff findings. Existing unrelated findings elsewhere in the touched legacy files were not changed.
- `py_compile` passes for all touched Python files.
- `git diff --check` passes.
- Independent final review found no blocking or medium findings.

## Type of Change

- [x] Bug fix (non-breaking)
- [ ] New feature (non-breaking)
- [ ] New framework / benchmark integration
- [ ] Breaking change
- [ ] Documentation only
- [ ] Refactor (no behavior change)

## Checklist

- [x] No unrelated changes mixed in
- [x] New features have example config in `examples/` — N/A (bug fix; no new feature)
- [x] No secrets, API keys, or large binaries committed
- [x] Files stay isolated — shared training/config files are modified only where required for this fix
- [x] No modifications to shared files, or justified above
- [x] If adding framework/benchmark: checkpoint uploaded to personal HF (public) — N/A (no framework/benchmark added)

## Breaking Changes

None.

## Screenshots / Logs (optional)

All 12 DeepSpeed combinations passed: VLA, VLM, and co-training under ZeRO-2/ZeRO-3 with one and two ranks. Each run verifies boundaries `[False, True]`, exactly one optimizer/global step, exactly one scheduler step, and the expected final parameter value (`0.95` for VLA/VLM and `0.90` for co-training).

### CODE REVIEW

**Comment by yeahjack** [comment]
@codex review

File: starVLA/config/deepseeds/ds_config.yaml [hunk 1]
```diff
@@ -7,7 +7,7 @@
     },
     "train_micro_batch_size_per_gpu": "auto",
     "train_batch_size": "auto",
-    "gradient_accumulation_steps": 1,
+    "gradient_accumulation_steps": "auto",
```
[the remark this answers is not in the document]
Comment (mindorigin150) [reply starVLA/config/deepseeds/ds_config.yaml:10]: Addressed in `902e95d`.

The wiring concern was valid: all three training entry points now construct `Accelerator` and `DeepSpeedPlugin` after the trainer config is loaded, and pass `cfg.trainer.gradient_accumulation_steps` to `Accelerator`. `DeepSpeedPlugin` intentionally remains unset so the DeepSpeed `"auto"` field resolves from Accelerator while preserving Accelerate's launch-environment precedence.

One correction to the failure mode: with the pinned Accelerate 1.5.2, `prepare()` fills an `"auto"` DeepSpeed accumulation value from `Accelerator.gradient_accumulation_steps`. Before this fix it therefore resolved to Accelerator's unintended default `1`; it was not left unresolved and did not fail initialization. The new commit makes the trainer YAML value the default source instead.

Verified with `python tests/test_gradient_accumulation_config.py`, syntax compilation of the three trainers and the new test, and `git diff --check`.

[the remark this answers is not in the document]
Comment (mindorigin150) [reply starVLA/config/deepseeds/ds_config.yaml:10]: Follow-up refactor in `5e78f13`: Accelerator/DeepSpeed construction now lives in one shared `starVLA.training.accelerator_utils.build_accelerator` helper, and all three trainer entry points call it after config loading. The focused regression test now targets the single shared implementation. Re-ran the test, syntax compilation, and `git diff --check`.

[the remark this answers is not in the document]
Comment (mindorigin150) [reply starVLA/config/deepseeds/ds_config.yaml:10]: Corrective cleanup in `9303f3e`: the shared `build_accelerator` implementation now lives in the existing `starVLA.training.trainer_utils.trainer_tools` module, which all three trainers already use. The temporary standalone `accelerator_utils.py` module has been removed. Tests, syntax compilation, and `git diff --check` pass.

[the remark this answers is not in the document]
Comment (mindorigin150) [reply starVLA/config/deepseeds/ds_config.yaml:10]: Finalized in `ccb6e023` after rebasing onto the current `starVLA_dev`. All three entrypoints now build Accelerator from the wrapped trainer config. Added deterministic CPU regression coverage plus real 1/2-rank ZeRO-2 and ZeRO-3 production-path smoke tests; configured GAS=2 resolves correctly, produces one optimizer/scheduler update, and both microbatches contribute to the expected parameter value.
